dashboard/app/controllers/script_levels_controller.rb

```diff
@@ -247,15 +247,35 @@ def stage_extras
       return
     end
 
-    @stage = Script.get_from_cache(params[:script_id]).lesson_by_relative_position(params[:stage_position].to_i)
+    user = @user || current_user
+    @stage = Script.get_from_cache(
+      params[:script_id]
+    ).lesson_by_relative_position(
+      params[:stage_position].to_i
+      )
     @script = @stage.script
+    script_bonus_levels_by_stage = @script.get_bonus_script_levels(@stage)
+
+    # get user progress for bonus levels
+    script_bonus_levels_by_stage.each do |stage|
+      stage[:levels].each do |level_summary|
+        user_level = UserLevel.find_by(
+          level_id: level_summary[:level_id], user_id: user.id, script: @script
+        )
+        level_summary[:perfect] = user_level ? true : false
```

Comment:
Not sure if this change was intentional or not, but the `true` case used to return `user_level&.perfect?`

Reply:
that call is redundant because `user_level != nil === user_level.perfect? == true`, but `level_summary[:perfect] = user_level&.perfect?` seems clearer, so i will update